File: py/classical/substitution.py
import string
from collections import Counter

# English letter frequency order (most → least common)
ENGLISH_FREQ = "ETAOINSHRDLCUMWFGYPBVKJXQZ"


def _build_maps(key: str):
    """Return (enc_map, dec_map) dicts from a 26-letter key string."""
    key = key.upper().replace(" ", "")
    if len(key) != 26 or len(set(key)) != 26:
        raise ValueError("Key must be a 26-letter permutation of the alphabet.")
    enc = {string.ascii_uppercase[i]: key[i] for i in range(26)}
    dec = {v: k for k, v in enc.items()}
    return enc, dec


def _apply_map(text: str, mp: dict) -> str:
    result = []
    for ch in text.upper():
        result.append(mp.get(ch, ch))
    return "".join(result)


def handle(operation: str, params: dict) -> dict:
    op = operation.lower()

    if op == "encrypt":
        plaintext = params.get("plaintext", "")
        key       = params.get("key", "")
        enc_map, _ = _build_maps(key)
        ciphertext = _apply_map(plaintext, enc_map)
        # Build readable key table (A→Q, B→W, …)
        key_table = "\n".join(
            f"{string.ascii_uppercase[i]} → {key.upper()[i]}"
            for i in range(26)
        )
        return {
            "ciphertext": ciphertext,
            "key_table":  key_table,
        }

    elif op == "decrypt":
        ciphertext = params.get("ciphertext", "")
        key        = params.get("key", "")
        _, dec_map = _build_maps(key)
        plaintext  = _apply_map(ciphertext, dec_map)
        return {"plaintext": plaintext}

    elif op == "brute":
        # Try all 26 Caesar-shift keys
        ciphertext = params.get("ciphertext", "").upper()
        results = []
        for shift in range(26):
            dec = {
                string.ascii_uppercase[(i + shift) % 26]: string.ascii_uppercase[i]
                for i in range(26)
            }
            candidate = _apply_map(ciphertext, dec)
            results.append(f"Shift {shift:2d}: {candidate}")
        return {"brute_force_results": "\n".join(results)}

    elif op == "frequency":
        ciphertext = params.get("ciphertext", "").upper()
        counts     = Counter(ch for ch in ciphertext if ch.isalpha())
        total      = sum(counts.values()) or 1

        # Sort ciphertext letters by frequency (most → least)
        sorted_cipher = [ch for ch, _ in counts.most_common()]

        # Map each cipher letter → guessed plaintext letter by English freq rank
        freq_map = {}
        for i, cipher_ch in enumerate(sorted_cipher):
            if i < len(ENGLISH_FREQ):
                freq_map[cipher_ch] = ENGLISH_FREQ[i]

        # Frequency table lines
        freq_table_lines = []
        for ch, cnt in counts.most_common():
            pct   = cnt / total * 100
            guess = freq_map.get(ch, "?")
            freq_table_lines.append(
                f"{ch}: {cnt:4d} ({pct:5.1f}%)  → guessed plaintext: {guess}"
            )

        # Suggested decryption using frequency mapping
        suggested = "".join(freq_map.get(ch, ch) for ch in ciphertext)

        return {
            "frequency_table":       "\n".join(freq_table_lines),
            "suggested_plaintext":   suggested,
            "cipher_to_plain_guess": str(freq_map),
        }

    else:
        raise ValueError(f"Unknown operation '{operation}' for substitution cipher.")

File: internal/engine/classical/substitution.go
// Package classical implements classical cryptographic algorithms in Go.
package classical

import (
	"fmt"
	"sort"
	"strings"
	"unicode"
)

// englishFreq is the standard English letter frequency order (most → least).
const englishFreq = "ETAOINSHRDLCUMWFGYPBVKJXQZ"

// Substitution handles encrypt / decrypt / frequency operations
// for the monoalphabetic substitution cipher.
func Substitution(operation string, params map[string]string) (map[string]string, error) {
	op := strings.ToLower(operation)

	switch op {
	case "encrypt":
		plaintext := strings.ToUpper(params["plaintext"])
		key := strings.ToUpper(strings.ReplaceAll(params["key"], " ", ""))
		if len(key) != 26 {
			return nil, fmt.Errorf("key must be exactly 26 letters, got %d", len(key))
		}
		encMap := buildEncMap(key)
		ciphertext := applyMap(plaintext, encMap)
		return map[string]string{
			"ciphertext": ciphertext,
			"key_table":  buildKeyTable(key),
		}, nil

	case "decrypt":
		ciphertext := strings.ToUpper(params["ciphertext"])
		key := strings.ToUpper(strings.ReplaceAll(params["key"], " ", ""))
		if len(key) != 26 {
			return nil, fmt.Errorf("key must be exactly 26 letters, got %d", len(key))
		}
		decMap := buildDecMap(key)
		return map[string]string{"plaintext": applyMap(ciphertext, decMap)}, nil

	case "frequency":
		ciphertext := strings.ToUpper(params["ciphertext"])
		counts := make(map[rune]int)
		total := 0
		for _, ch := range ciphertext {
			if ch >= 'A' && ch <= 'Z' {
				counts[ch]++
				total++
			}
		}
		if total == 0 {
			return nil, fmt.Errorf("no alphabetic characters in ciphertext")
		}

		type pair struct {
			ch  rune
			cnt int
		}
		var pairs []pair
		for ch, cnt := range counts {
			pairs = append(pairs, pair{ch, cnt})
		}
		sort.Slice(pairs, func(i, j int) bool { return pairs[i].cnt > pairs[j].cnt })

		freqMap := make(map[rune]rune)
		for i, p := range pairs {
			if i < len(englishFreq) {
				freqMap[p.ch] = rune(englishFreq[i])
			}
		}

		var tableLines, suggested strings.Builder
		for _, p := range pairs {
			pct := float64(p.cnt) / float64(total) * 100
			guess, _ := freqMap[p.ch]
			tableLines.WriteString(fmt.Sprintf("%c: %4d (%5.1f%%)  → guessed plaintext: %c\n", p.ch, p.cnt, pct, guess))
		}
		for _, ch := range ciphertext {
			if g, ok := freqMap[ch]; ok {
				suggested.WriteRune(g)
			} else if unicode.IsLetter(ch) {
				suggested.WriteRune(ch)
			} else {
				suggested.WriteRune(ch)
			}
		}
		return map[string]string{
			"frequency_table":     strings.TrimRight(tableLines.String(), "\n"),
			"suggested_plaintext": suggested.String(),
		}, nil

	default:
		return nil, fmt.Errorf("unknown operation %q for substitution cipher", operation)
	}
}

func buildEncMap(key string) map[rune]rune {
	m := make(map[rune]rune, 26)
	for i, ch := range key {
		m[rune('A'+i)] = ch
	}
	return m
}

func buildDecMap(key string) map[rune]rune {
	m := make(map[rune]rune, 26)
	for i, ch := range key {
		m[ch] = rune('A' + i)
	}
	return m
}

func applyMap(text string, mp map[rune]rune) string {
	var sb strings.Builder
	for _, ch := range text {
		if mapped, ok := mp[ch]; ok {
			sb.WriteRune(mapped)
		} else {
			sb.WriteRune(ch)
		}
	}
	return sb.String()
}

func buildKeyTable(key string) string {
	var sb strings.Builder
	for i, ch := range key {
		sb.WriteString(fmt.Sprintf("%c → %c\n", 'A'+i, ch))
	}
	return strings.TrimRight(sb.String(), "\n")
}
